contrib/go/src/python/pants/contrib/go/tasks/go_binary_fingerprint_strategy.py

```diff
@@ -32,7 +32,7 @@ def compute_fingerprint(self, target):
       return fp
 
     hasher = hashlib.sha1()
-    hasher.update(fp)
+    hasher.update(fp.encode('utf-8'))
```

Comment:
It would be very neat if we could allow `FingerprintStrategy` implementations to avoid having to concern themselves with encodings, this feels like something that might surprise pants users with plugins in their own repo when they switch to pants with python 3 (I'm selfishly concerned about our own internal repo). Is there a way we could avoid this? Asking to see if we can turn it into a followup issue or PR, not blocking for this one.

Reply:
Do you mean not concern themselves with `utf-8` vs. other encodings like `ascii`, or with having to encode to bytes to begin with? 

The issue is that `hashlib.sha1()` requires bytes no matter what, and chokes on unicode. We consistently use UTF-8 everywhere in Pants, and enforce that users are running with UTF-8 (https://github.com/pantsbuild/pants/blob/master/src/python/pants/bin/pants_loader.py#L44), so I think it is safe for us to consistently encode to UTF-8 with FingerprintStrategies. 

It's also a pattern we use in I think now 100% of the Fingerprint instances we have. These two files appear to have been the last places I had not updated.

Reply:
> with having to encode to bytes to begin with?

Mostly this, although I'm not clear on whether being explicit here is pythonic -- it feels error-prone.

> It's also a pattern we use in I think now 100% of the Fingerprint instances we have.

But is it still true that users of pants in other repos who want to upgrade to python 3 pants will have to make this change in every one of their `FingerprintStrategy` implementations? How can we make sure we do this correctly?

Reply:
I didn't realize people implement their own FingerprintStrategy? They will have to add an `.encode('utf-8')` anywhere they currently are passing Unicode. (They can actually encode with any encoding they want, so long as the result is type `bytes`)

The error message from Python 3 is quite explicit:
```python
from hashlib import sha1

hasher = sha1()
hasher.update("😜")
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-a47404780295> in <module>
----> 1 hasher.update("😜")

TypeError: Unicode-objects must be encoded before hashing
```

I also think it could be helpful for us to develop a concise migration guide going over issues like this for issues people are likely to run into! I'd be happy to work on that with people.

--

I do not think the right solution is us encoding things for them. Explicit is better than implicit, which is the whole point of Python 3 not allowing unicode here.

I'm also not sure we could even fix the compat issues for them even if we wanted to. Every implementation of `FingerprintStrategy` I've seen that does not use `stable_json_hash` explicitly calls `hashlib.sha1()`, and the issue is with the local definition of `compute_fingerprint()`, not our Pants logic. That is, the `TypeError` would surface before even hitting our code. The only workaround I'd see is catching `TypeError`, which feels very wrong.